primitives/state-machine/src/lib.rs

```diff
@@ -209,94 +231,168 @@ mod execution {
 	#[derive(Debug, Clone)]
 	pub enum BackendTrustLevel {
 		/// Panics from trusted backends are considered justified, and never caught.
 		Trusted,
-		/// Panics from untrusted backend are caught and interpreted as runtime error.
-		/// Untrusted backend may be missing some parts of the trie, so panics are not considered
-		/// fatal.
+		/// Panics from untrusted backend are caught and interpreted as runtime error. Untrusted
+		/// backend may be missing some parts of the trie, so panics are not considered fatal.
 		Untrusted,
 	}
 
-	/// Like `ExecutionStrategy` only it also stores a handler in case of consensus failure.
+	/// Like `ExecutionStrategy`, only it also stores a handler in case of consensus failure.
 	#[derive(Clone)]
-	pub enum ExecutionManager<F> {
+	pub enum ExecutionStrategyWithHandler<F> {
 		/// Execute with the native equivalent if it is compatible with the given wasm module;
 		/// otherwise fall back to the wasm.
 		NativeWhenPossible,
-		/// Use the given wasm module. The backend on which code is executed code could be
-		/// trusted to provide all storage or not (i.e. the light client cannot be trusted to provide
-		/// for all storage queries since the storage entries it has come from an external node).
+		/// Use the given wasm module. The backend on which code is executed code could be trusted
+		/// to provide all storage or not (i.e. the light client cannot be trusted to provide for
+		/// all storage queries since the storage entries it has come from an external node).
 		AlwaysWasm(BackendTrustLevel),
-		/// Run with both the wasm and the native variant (if compatible). Call `F` in the case of any discrepancy.
+		/// Run with both the wasm and the native variant (if compatible). Call `F` in the case of
+		/// any discrepancy.
 		Both(F),
 		/// First native, then if that fails or is not possible, wasm.
 		NativeElseWasm,
 	}
 
-	impl<'a, F> From<&'a ExecutionManager<F>> for ExecutionStrategy {
-		fn from(s: &'a ExecutionManager<F>) -> Self {
-			match *s {
-				ExecutionManager::NativeWhenPossible => ExecutionStrategy::NativeWhenPossible,
-				ExecutionManager::AlwaysWasm(_) => ExecutionStrategy::AlwaysWasm,
-				ExecutionManager::NativeElseWasm => ExecutionStrategy::NativeElseWasm,
-				ExecutionManager::Both(_) => ExecutionStrategy::Both,
-			}
+	/// The execution manager.
+	///
+	/// Same as [`ExecutionConfig`], but the inner `strategy` contains failure handler as well.
+	#[derive(Clone)]
+	pub struct ExecutionManager<F> {
+		/// The strategy of the execution's runtime.
+		pub strategy: ExecutionStrategyWithHandler<F>,
+		/// The context in which the aforementioned strategy is being used.
+		pub context: ExecutionContext,
+	}
+
+	#[cfg(test)]
+	impl<F> ExecutionManager<F> {
+		/// Create a new instance of [`Self`] from the given strategy for testing.
+		pub(crate) fn new(strategy: ExecutionStrategyWithHandler<F>) -> Self {
+			Self { strategy, context: ExecutionContext::Offchain }
```

Comment:
I would generally use the `Consensus` context as default, and have a specific `new_offchain` when we need to do something specific.
(not sure where this function is used but could end being a footgun).

Reply:
This is only used for testing, but yeah I changed it anyhow.